Retry birger commands

JAAlvarado-Montes opened the pull request:
Retry the command until it matches the birger response.

## Description
The command is sent many time until both the command and the response from the birger are the same.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** document.

AnthonyHorton commented on src/panoptes/pocs/focuser/birger.py:
```diff
@@ -249,13 +249,15 @@ def _send_command(self, command, response_length=None):
         self._serial_port.reset_input_buffer()
 
-        # Send command
-        self._serial_io.write(command + '\r')
```
You shouldn't take this out, otherwise you're trying to read the response before you've even sent the command once. This will cause a delay equal to the serial port timeout for no reason, every single time we send a command.

JAAlvarado-Montes replied:
Got it! Didn't realize that, I thought that reading without sending the command we would have gotten an `''`, so it'd get into the while loop and send the first command from there

JAAlvarado-Montes replied:
What I'm going to do is to move everything inside the loop, even the first time that the command is sent. That way I wouldn't be reading before sending the command for the first time.

AnthonyHorton commented on src/panoptes/pocs/focuser/birger.py:
```diff
@@ -249,13 +249,15 @@ def _send_command(self, command, response_length=None):
-        assert echo == command, self.logger.warning("echo != command: {} != {}".format(
-            echo, command))
+        echo = [self._serial_io.readline().rstrip()]
+        while echo[0] != command:
+            # Send command
+            self._serial_io.write(command + '\r')
+            echo[0] = self._serial_io.readline().rstrip()
```
If the Birger has a problem worse than intermittent serial comms error then this will be an infinite loop! You should _always_ have a finite number of retries, with a small default value, and a corresponding config parameter to change it.

If the expected response is not received within the set number of retries then the code should break out of the loop, log an error, and throw an exception.

AnthonyHorton commented on src/panoptes/pocs/focuser/birger.py:
```diff
@@ -249,13 +249,15 @@ def _send_command(self, command, response_length=None):
-        assert echo == command, self.logger.warning("echo != command: {} != {}".format(
-            echo, command))
+        echo = [self._serial_io.readline().rstrip()]
+        while echo[0] != command:
```
Should log a warning each time the expect response is not received and a retry is needed. We want a record of how often this is happening.

AnthonyHorton commented on src/panoptes/pocs/focuser/birger.py:
```diff
@@ -249,13 +249,15 @@ def _send_command(self, command, response_length=None):
+            self._serial_io.write(command + '\r')
+            echo[0] = self._serial_io.readline().rstrip()
+
+        assert echo[0] == command, self.logger.warning("echo != command: {} != {}".format(
+            echo[0], command))
```
They way you currently have the loop working this is totally redundant code, as there is no way get out of the loop until `echo[0] == command` is `True`.

AnthonyHorton commented on src/panoptes/pocs/focuser/birger.py:
```diff
@@ -249,13 +249,15 @@ def _send_command(self, command, response_length=None):
+            echo[0], command))
 
         # Adaptor should then send 'OK', even if there was an error.
         ok = self._serial_io.readline().rstrip()
```
Note that there are several more steps to verifying responses from the Birger. Checking that the Birger echos the command that was sent to it is just the first step.

We also check here that this is followed by 'OK', check for error codes, then retrieve any additional lines of response that are expected for the given command type.

The serial comms problems are just as likely to effect these subsequent checks as the initial echo check, so they should all be moved inside the retry loop.

JAAlvarado-Montes replied:
But what if the response from the birger is correct when the command is sent for the first time as expected? If everything is inside the retry loop (which only runs if `echo != command`), that'd mean there would not be subsequent checks if the loop is not triggered.

AnthonyHorton reviewed (changes requested):
It's the right idea, but I have several objections to the way you've implemented it.

JAAlvarado-Montes commented:
Hi @AnthonyHorton,

Please review again this PR when you've got a chance. I have addressed all of your comments, though there might still be some things to improve. I would really appreciate your input. Thank you!

Cheers,
Jaime A.

danjampro commented on src/panoptes/pocs/focuser/birger.py:
```diff
@@ -63,10 +63,13 @@ def __init__(self,
         super().__init__(name=name, model=model, *args, **kwargs)
         self.logger.debug('Initialising Birger focuser')
 
+        self.max_command_attempts = max_command_attempts
```
@JAAlvarado-Montes I thinks this should be a private variable, `self._max_command_attempts`

danjampro replied:
I think this line needs to come before calling `super().__init__`. Right now I'm getting this error on creating a focuser:
```
D 01-12 03:59:24.850 (03:59:24 UTC) | panoptes.pocs.camera.camera _create_subcomponent:1039        | Creating the base_class_name='AbstractFocuser' object from dict
D 01-12 03:59:24.857 (03:59:24 UTC) | panoptes.pocs.focuser.focuser __init__:99                    | Focuser created: Birger Focuser on /dev/ttyUSB0
D 01-12 03:59:24.873 (03:59:24 UTC) | panoptes.pocs.focuser.serial _connect:107                    | Established serial connection to Birger Focuser on /dev/ttyUSB0.
E 01-12 03:59:24.879 (03:59:24 UTC) | huntsman.pocs.utils.pyro.cli service:87                      | Pyro huntsmanpi000 shutdown unexpectedly AttributeError("'Focuser' object has no attribute 'max_command_attempts'")
```

danjampro reviewed (approved):
LGTM and tested on Huntsman

wtgee commented on src/panoptes/pocs/focuser/birger.py:
```diff
@@ -249,36 +253,50 @@ def _send_command(self, command, response_length=None):
+            # Adaptor should then send 'OK', even if there was an error.
+            ok = self._serial_io.readline().rstrip()
+
+            if echo == command and ok == 'OK':
+                if response_length == 0:
+                    # Not expecting any further response. Should check the buffer anyway in case an error
+                    # message has been sent.
+                    if self._serial_port.in_waiting:
+                        response.append(self._serial_io.readline())
+
+                elif response_length > 0:
+                    # Expecting some number of lines of response. Attempt to read that many lines.
+                    for i in range(response_length):
+                        response.append(self._serial_io.readline())
+
+                else:
+                    # Don't know what to expect. Call readlines() to get whatever is there.
+                    response.append(self._serial_io.readlines())
```
The conditionals here don't do anything.  In all situations you are just reading the buffer and storing it. There shouldn't be anything coming back after the `response_length` and if so maybe that shouldn't be added to the `response` anyway.

```suggestion
            if echo == command and ok == 'OK':
                # Get any available response
                response.append(self._serial_io.readlines())
```

wtgee commented on src/panoptes/pocs/focuser/birger.py:
```diff
@@ -249,36 +253,50 @@ def _send_command(self, command, response_length=None):
+                break
+
+            if echo != command:
+                self.logger.warning(f"echo != command: {echo} != {command}. Retrying command.")
```
```suggestion
                self.logger.warning(f"echo != command: {echo!r} != {command!r}. Retrying command.")
```

wtgee commented on src/panoptes/pocs/focuser/birger.py:
```diff
@@ -249,36 +253,50 @@ def _send_command(self, command, response_length=None):
+                self.logger.warning(f"echo != command: {echo} != {command}. Retrying command.")
+
+            if ok != 'OK':
+                self.logger.warning(f"ok != 'OK': {ok} != 'OK'. Retrying command.")
```
Do we know in what situations the `echo` would be correct but the `ok` wouldn't?

```suggestion
                self.logger.warning(f"ok != 'OK': {ok!r} != 'OK'. Retrying command.")
```

wtgee commented on src/panoptes/pocs/focuser/birger.py:
```diff
@@ -249,36 +253,50 @@ def _send_command(self, command, response_length=None):
+            if ok != 'OK':
+                self.logger.warning(f"ok != 'OK': {ok} != 'OK'. Retrying command.")
+
+        try:
+            assert (echo, ok) == (command, 'OK')
+        except AssertionError as err:
+            self.logger.error(err)
+            warn(err)
```
Don't you do all of this above?

```suggestion
```

wtgee reviewed (changes requested):
Seems like this should be built into the `AbstractSerialFocuser` rather than just the Birger `Focuser`.  Any reason it's not?

JAAlvarado-Montes commented:
Hi @wtgee , this is done only for birgers because only these controllers send a response back that needs to be verified with the command sent. The astromechs don't do this.

I've addressed all of your comments above 👍 

Cheers,
Jaime A.

wtgee commented:
> Hi @wtgee , this is done only for birgers because only these controllers send a response back that needs to be verified with the command sent. The astromechs don't do this.

Well, there are two things going on in this PR, namely 1) retrying commands that fail and 2) verifying the response.

The retrying of the commands (the title of the PR) should be generic and available to all serial-based focusers.  The verifying of the response could potentially be generic as well.

JAAlvarado-Montes commented:
However, we don't need to retry commands for other focusers, that's only needed for the birgers. Each focuser has its own way to read the commands, that's why `_send_command` is different for each focuser controller.

AnthonyHorton commented on src/panoptes/pocs/focuser/birger.py:
```diff
@@ -249,36 +253,32 @@ def _send_command(self, command, response_length=None):
         # Clear the input buffer in case there's anything left over in there.
         self._serial_port.reset_input_buffer()
```
This should be done before each time you send the command, i.e. you should move this line inside the loop just before the `self._serial_io.write(command + '\r')` line.

If you don't clear the buffer each time you might end up reading part of the response from a previous failed attempt to send the command, causing the response validation to fail unnecessarily.

AnthonyHorton commented on src/panoptes/pocs/focuser/birger.py:
```diff
@@ -249,36 +253,32 @@ def _send_command(self, command, response_length=None):
+
+            if echo == command and ok == 'OK':
+                # Get any avaialble response left in buffer.
+                response.append(self._serial_io.readlines())
```
You've removed some of the functionality of the original code here. Almost all of the Birger commands have an expected response length, i.e. how many lines of response there should be after the echo and the 'OK'. 

`self._serial_io.readlines()` is supposed to be the fall back option only used for the 1 or 2 commands that return an unknown number of lines of response because it will block for the duration of the serial timeout waiting for more lines of response. When the number of lines is known it is far quicker to read only the expected number of lines.

We always try to read at least 1 line though, because it's possible there will be an error message immediately after the 'OK'.

```suggestion
                if response_length == 0:
                    # Not expecting any further response. Should check the buffer anyway in case an error
￼                   # message has been sent.
                    response.append(self._serial_io.readline())
                elif response_length > 0:
                    # Expecting some number of lines of response. Attempt to read that many lines.
                    for i in range(response_length):
￼                        response.append(self._serial_io.readline())
                else:
                    # Don't know what to expect. Call readlines() to get whatever is there.
                    response.append(self._serial_io.readlines())
```

AnthonyHorton commented on src/panoptes/pocs/focuser/birger.py:
```diff
@@ -249,36 +253,32 @@ def _send_command(self, command, response_length=None):
+                # Get any avaialble response left in buffer.
+                response.append(self._serial_io.readlines())
+
+                self.logger.debug(f"Got correct response after {i + 1} attempts")
```
It's debatable but I feel that this message is unnecessary, or at least should be downgraded from DEBUG to TRACE level. If retries are needed then they will be logged and this message won't add much info, and if it works first time like it's supposed to we probably don't need a additional debug message either.

AnthonyHorton commented on src/panoptes/pocs/focuser/birger.py:
```diff
@@ -249,36 +253,32 @@ def _send_command(self, command, response_length=None):
+                self.logger.debug(f"Got correct response after {i + 1} attempts")
+                break
+
+            if echo != command:
+                self.logger.warning(f"echo != command: {echo!r} != {command!r}. Retrying command.")
+
+            if ok != 'OK':
+                self.logger.warning(f"ok != 'OK': {ok!r} != 'OK'. Retrying command.")
```
You are checking the echo and the 'OK' inside the loop, but you've left the check for an error message outside the loop. I know that our bad serial connections are (so far) causing things to fail at the echo check but it is possible that other intermittent Birger problems could pass the echo and 'OK' check but cause occasional error messages to occur (e.g. ERR21) and the retry code could work around those too if you move the check inside the loop.

AnthonyHorton commented on src/panoptes/pocs/focuser/birger.py:
```diff
@@ -249,36 +253,32 @@ def _send_command(self, command, response_length=None):
+
+            if ok != 'OK':
+                self.logger.warning(f"ok != 'OK': {ok!r} != 'OK'. Retrying command.")
 
```
Important thing missing here: there is no check that the command eventually succeeds! If the checks fail it will keep retrying until the maximum number of retries have been reached, but after that the code just carries on regardless.

You need something like a `success` variable that get set to `False` at the start, and gets set to `True` if all the checks pass. Once you exit the loop you need to check the `success` variable and if it is not `True` you need to raise a `PanError` with a suitable message.

AnthonyHorton reviewed (changes requested):
Several things I'd still like to see changed here, but the show stopper is the absence of a check that the command eventually succeeds. At the moment the code will just carry on regardless once the maximum number of retries has been reached, even if all attempts fail.

AnthonyHorton commented:
> Regarding 1) we don't need to retry commands for other focusers, that's only needed for the birgers. Each focuser has its own way to read the commands that are sent to them, that's why `_send_command` is different for each focuser controller. Moreover, it's only the birgers which send a response back and the `OK` verification (which are the two things we use to see if a command needs to be retried). It'd be completely unnecessary and a mistake having all of this for other focuser (e.g. the astromechs).

You can still make this generic. There is no particular reason why we should only retry commands for the Birgers. Any serial connection can drop characters if there is a bad connection, all the focusers we support use serial connections, and all the focusers send a checkable response to at least some of their commands, so all of them could potentially benefit from the retry functionality. The way to implement this would be to have a generic `send_command()` method in the serial focuser base class which includes the retry functionality, and which uses calls to a focuser specific `_validate_response()` method which reads the response (if any) & returns, say, a tuple of `(success, response)`.

> Regarding 2), the verifying response was already there and it's always been only for the birgers focuser, I only moved it into the retry loop. We don't this for the other focusers.

It was written like that because at that point the Birgers were the only focuser. Now there are 3 types we have a better idea of what functionality they have in common (or should have in common), and that common functionality is better implemented in the base class.

AnthonyHorton commented on src/panoptes/pocs/focuser/birger.py:
```diff
@@ -246,39 +250,62 @@ def _send_command(self, command, response_length=None):
-                response.append(self._serial_io.readline())
+            if echo == command and ok == 'OK':
+                if response_length == 0:
+                    # Not expecting any further response. Should check the buffer anyway in case an error
+                    # message has been sent.
+                    if self._serial_port.in_waiting:
+                        response.append(self._serial_io.readline())
 
-        else:
-            # Don't know what to expect. Call readlines() to get whatever is there.
-            response.append(self._serial_io.readlines())
+                elif response_length > 0:
+                    # Expecting some number of lines of response. Attempt to read that many lines.
+                    for i in range(response_length):
+                        response.append(self._serial_io.readline())
+
+                else:
+                    # Don't know what to expect. Call readlines() to get whatever is there.
+                    response.append(self._serial_io.readlines())
+
+                try:
+                    assert (echo, ok) == (command, 'OK')
+                    success = True
+                except AssertionError as err:
+                    self.logger.error(err)
+                    warn(err)
+
+                self.logger.trace(f"Got correct response after {i + 1} attempts")
+                break
+
+            if echo != command:
+                self.logger.warning(f"echo != command: {echo!r} != {command!r}. Retrying command.")
+
+            if ok != 'OK':
+                self.logger.warning(f"ok != 'OK': {ok!r} != 'OK'. Retrying command.")
+
+        try:
+            assert(success)
+        except AssertionError:
+            raise error.PanError(f'The command {command!r} was not properly read by {self}')
```
With all the edits this has ended up a bit messy. Suggestion to clean it up:

```
        response = []
        success = False

        for i in range(self._max_command_retries):
            self._serial_port.reset_input_buffer()
            self._serial_io.write(command + '\r')

            echo = self._serial_io.readline().rstrip()
            if echo != command:
                self.logger.warning(f"echo != command: {echo!r} != {command!r}. Retrying command.")
                continue

            ok = self._serial_io.readline().rstrip()
            if ok != 'OK':
                self.logger.warning(f"ok != 'OK': {ok!r} != 'OK'. Retrying command.")
                continue

            if response_length == 0:
                # Not expecting any further response. Should check the buffer anyway in case an error
                # message has been sent.
                if self._serial_port.in_waiting:
                    response.append(self._serial_io.readline())
            elif response_length > 0:
                # Expecting some number of lines of response. Attempt to read that many lines.
                for i in range(response_length):
                     response.append(self._serial_io.readline())
            else:
                # Don't know what to expect. Call readlines() to get whatever is there.
                response.append(self._serial_io.readlines())

            success = True
            break

        if not success:
            raise error.PanError(f'Failed command {command!r} on {self}')

```

AnthonyHorton reviewed (changes requested):
With all the back and forth edits this has ended up a bit messy. I have suggested a cleaner rewrite of the main part of the code.

In addition this is still only checking for error codes after the retry loop, instead of inside the loop, which would be better.

wtgee reviewed (approved):
Thanks @JAAlvarado-Montes and @AnthonyHorton. Approving here to clear my change request, will merge on passing tests.

AnthonyHorton reviewed (approved):
LGTM.
